### PR TITLE
gdal_edit: add -a_coord_epoch option and update doc

### PR DESCRIPTION
**Pull request by dshean**
## What does this PR do?

Add new options to `gdal_edit.py` that will add, update or remove a user-defined coordinate epoch from the dataset SRS.  Updated documentation.  

## What are related issues/pull requests?

https://lists.osgeo.org/pipermail/gdal-dev/2024-January/058314.html

## Tasklist

 - [x] Make sure code is correctly formatted (cf [pre-commit configuration](https://gdal.org/development/dev_practices.html#commit-hooks))
 - [x] Add test case(s)
 - [x] Add documentation
 - [x] Updated Python API documentation (swig/include/python/docs/)
 - [ ] Review
 - [x] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

### CODE REVIEW

**Comment by dshean** [comment]
I don't think there are any tests for `gdal_edit.py`.  I tested several sequences and combinations of `-a_srs`, `-a_coord_epoch` and `-unsetepoch` for a sample GeoTiff.

I added the `-unsetepoch` option to maintain consistency with other existing options, like `-unsetnodata`.  It would be simpler to use `-a_coord_epoch 0.0` to unset, as that is what is returned by GetCoordinateEpoch() when the SRS has no epoch defined.  I expect this will evolve as support for epoch-aware CRS transformations evolves.

No changes to Python API (or associated doc).

**Comment by rouault** [comment]
Thanks! Tests for gdal_edit are in autotest/pyscripts/test_gdal_edit.py

File: doc/source/programs/gdal_edit.rst [hunk 1]
```diff
@@ -58,6 +59,18 @@ It works only with raster formats that support update access to existing dataset
     coordinate system will be removed (for TIFF/GeoTIFF, might not be well
     supported besides that).
 
+.. option:: -a_coord_epoch <epoch>
+
+    Assign/override the coordinate epoch of the dataset, as a decimal year (e.g., 2021.3).
+
+    .. versionadded:: 3.4
```
Comment (rouault) [inline doc/source/programs/gdal_edit.rst:66]:
```suggestion
    .. versionadded:: 3.9
```

File: doc/source/programs/gdal_edit.rst [hunk 2]
```diff
@@ -58,6 +59,18 @@ It works only with raster formats that support update access to existing dataset
...
+
+.. option:: -unsetepoch
+
+    Remove the coordinate epoch of the dataset.
+
+    .. versionadded:: 3.4
```
Comment (rouault) [inline doc/source/programs/gdal_edit.rst:72]:
```suggestion
    .. versionadded:: 3.9
```

**Comment by dshean** [comment]
OK, I added a simple test and updated the version number to 3.9 in the doc.

I did a rebase instead of merge before pushing changes.  Let me know if I should clean up, or if you can isolate the relevant commit [0d5fa29](https://github.com/OSGeo/gdal/pull/9144/commits/0d5fa293fdc2c3a079c9066fba7f864496fd724c)

**Comment by rouault** [comment]
> I did a rebase instead of merge before pushing changes.

actually, you did the reverse ;-) I'd prefer indeed that it is a rebase with a single commit

**Comment by dshean** [comment]
Sorry about that - I'm a little rusty. I reset my local branch, squashed the two relevant commits, and did a force push to my remote branch, which I think takes care of the issue.  Should now be a single commit, I think ready to merge.  Thanks!

**Comment by rouault** [comment]
> Sorry about that - I'm a little rusty. I reset my local branch, squashed the two relevant commits, and did a force push to my remote branch, which I think takes care of the issue. Should now be a single commit, I think ready to merge. Thanks!

no problem. that looks good. I'm letting now CI run and see if things are in order

**Comment by dshean** [comment]
One other thought - I actually like the FlatGeoBuf strategy to include `EPOCH[2016.47]` in WKT2 (https://gdal.org/user/coordinate_epoch.html#flatgeobuf).  Easy to include this when defining custom WKT2 CRS for a particular dataset, and not limited to GeoTiff format.  Any plans to support that approach in PROJ/GDAL?

Also, it might be useful to have a centralized function to convert a datetime/Timestamp to decimal year for the epoch specification.  I have my own implementation of this, and I recall seeing something in the GDAL code, but I don't think it's exposed in the Python API.

**Comment by rouault** [comment]
> Any plans to support that approach in PROJ/GDAL?

well PROJ  (9.2 maybe, not sure exactly which version, but "recently") does support the COORDINATEMETADATA[] WKT support, and when this is ingested by GDAL, it feeds the EPOCH[] to the relevant member of OGRSpatialReference

> Also, it might be useful to have a centralized function to convert a datetime/Timestamp to decimal year for the epoch specification. 

I'm not sure about that. I've always seen the epoch being specified as a decimal year. I imagine that someone does the conversion, but it seems this is something generally done "before" PROJ or GDAL, mostly be geodesists

**Comment by dshean** [comment]
> I've always seen the epoch being specified as a decimal year. I imagine that someone does the conversion, but it seems this is something generally done "before" PROJ or GDAL, mostly be geodesists

Yes, epoch should always be specified as decimal year, before updating metadata in PROJ/GDAL.  

I'm just thinking about best place for a convenience function to convert the more common (YYYY, MM, DD) date information to a decimal year, since that's typically what is available from filename or metadata for a dataset.  

Most people who think about these issues have an existing capability to do the conversion to decimal year, usually with their own implementation (something like this: https://stackoverflow.com/a/6451892).  

But it presents an additional hurdle for newer users who want to perform epoch-aware CRS transformations, but only have the more common YYYYMMDD timestamp of their dataset from metadata, and don't have an existing utility to convert.

Not suggesting any change is necessary - users can figure this out on their own.  Alternative might be to add an option for a utility like gdal_edit to accept a timestamp (with known formatting) and then do the conversion to decimal year before setting the epoch.  Or maybe the conversion to decimal year could be easily added to Python datetime module or pandas.Timestamp.

**Pull request merged**